Automattic-Tracks-iOS/Model/ExponentialBackoffTimer.swift

```diff
@@ -0,0 +1,58 @@
+import Foundation
+
+struct ExponentialBackoffTimer {
+
+    /// The current amount of delay, in seconds
+    private(set) var delay: Int
+
+    /// The current exponent
+    private var exponent: Int = 1
+
+    /// Keep a copy of the initial delay to allow reset
+    private let initialDelay: Int = 0
+
+    /// The minimum allowed delay
+    let minimumDelay: Int
+
+    /// The maximum allowed delay
+    let maximumDelay: Int
+
+    /// Create a backoff timer with an optional initial duration.
+    ///
+    /// - Parameters:
+    ///   - minimumDelay: The smallest possible delay (specified in seconds). Must be greater than one.
+    ///   - maximumDelay: The largest possible delay (specified in seconds). Must be greater than one, and greater than the initial delay.
+    ///            The default is approximately one day.
+    init(minimumDelay: Int = 2, maximumDelay: Int = 86_400) {
+        precondition(minimumDelay > 1, "The initial delay must be greater than 1 – otherwise it will never increase")
+        precondition(maximumDelay > minimumDelay, "The limit must be greater than the initial delay")
+        precondition(maximumDelay < Int32.max / 2, "The limit must be less than Int32.max / 2 to avoid overflow")
+        self.delay = initialDelay
+        self.minimumDelay = minimumDelay
+        self.maximumDelay = maximumDelay
+    }
+
+    /// Exponentially increase the delay (up to `maximumDelay`)
+    mutating func increment() {
+        delay = Int(pow(Double(minimumDelay), Double(exponent)))
+        exponent += 1
+
+        if delay > maximumDelay {
+            delay = maximumDelay
+        }
+
+        next = .now() + .seconds(delay)
+        nextDate = Date(timeIntervalSinceNow: TimeInterval(delay))
+    }
+
+    /// Reset the delay to zero
+    mutating func reset() {
+        delay = initialDelay
+    }
+
+    /// A `DispatchTime` compatible with `DispatchQueue.asyncAfter` that represents the next time the timer should fire.
+    var next: DispatchTime = .now()
+
+    /// A `Date` representation of `next`.
+    var nextDate: Date = Date()
```

Comment:
You're spot on about the setters! Those are hidden as of #hashid

I would love to have a single source of truth for this, but unfortunately `DispatchTime` is backed by `dispatch_time_t`, which has this definition:

```
/**
 * @typedef dispatch_time_t
 *
 * @abstract
 * A somewhat abstract representation of time; where zero means "now" and
 * DISPATCH_TIME_FOREVER means "infinity" and every value in between is an
 * opaque encoding.
 */
public typealias dispatch_time_t = UInt64
```

Because there's no way to convert between them, this seemed like the cleanest method. The `next` is a machine-readable version of the value (used for scheduling the async task using `DispatchQueue` and `nextDate` is a human-readable version of the value (used for logging, debugging, or in a UI)